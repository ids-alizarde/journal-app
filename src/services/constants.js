export const Constants = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',

    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set active note',
    notesLoad: '[Notes] Load note',
    notesUpdated: '[Notes] Updated note',
    notesFileUrl: '[Notes] Update image URL',
    notesDelete: '[Notes] Delete note',
    notesLogOutCleaning: '[Notes] LogOut Cleaning',

    fireBaseTestEnvironment: 'test',
}