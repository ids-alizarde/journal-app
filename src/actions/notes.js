import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { Constants } from '../services/constants';
import { fileUpload } from '../services/fileUpload';
import { loadNotes } from '../services/loadNotes';


export const startNewNote = () => {
    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const doc = await db.collection(`${uid}/journal/notes`).add( newNote );

        dispatch( activeNote( doc.id, newNote ) );
        dispatch( addNewNote( doc.id, newNote ) );
    }
}

export const activeNote = ( id, note ) => ({

    type: Constants.notesActive,
    payload: {
        id,
        ...note
    }
})

export const addNewNote = ( id, note ) => ({
    type: Constants.notesAddNew,
    payload: {
        id, ...note 
    }
})

export const startLoadingNotes = ( uid ) => {
    return async ( dispatch ) => {

        const notes = await loadNotes( uid );
        dispatch( setNotes( notes) );
    }
}

export const setNotes = ( notes ) => ({

    type: Constants.notesLoad,
    payload: notes
})

export const startSaveNote = ( note ) => {
    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid;

        if ( !note.url ) {
            
            delete note.url;
        }

        const noteToFireStore = { ...note };

        delete noteToFireStore.id;

        await db.doc( `${ uid }/journal/notes/${ note.id }` ).update( noteToFireStore );

        dispatch( refreshNotes( note.id, noteToFireStore ) );

        Swal.fire( 'Saved', note.title, 'success');
    }
}

export const refreshNotes = ( id, note ) => ({
    type: Constants.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = ( file ) => {
    return async ( dispatch, getState ) => {

        const { active: activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading.......',
            text: 'Please wait........',
            allowOutsideClick: false,
            onBeforeOpen: () => {
                Swal.showLoading();
            }
        });

        const fileURL = await fileUpload( file );
        activeNote.url = fileURL;

        dispatch( startSaveNote( activeNote ) );

        Swal.close();
    }
}

export const startDeletingNote = ( id ) => {
    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid;
        await db.doc(`${ uid }/journal/notes/${ id }`).delete();
        dispatch( deleteNote( id ) );
    }
}

export const deleteNote = ( id ) => ({
    type: Constants.notesDelete,
    payload: id 
})

export const noteLogoutClean = () => ({
    type: Constants.notesLogOutCleaning
})