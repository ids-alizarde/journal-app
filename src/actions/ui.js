import { Constants } from "../services/constants";

export const setError = ( err ) => ({
    type: Constants.uiSetError,
    payload: err
});

export const removeError = ( err ) => ({
    type: Constants.uiRemoveError
});

export const startLoading = () => ({
    type: Constants.uiStartLoading
});

export const finishLoading = () => ({
    type: Constants.uiFinishLoading
});