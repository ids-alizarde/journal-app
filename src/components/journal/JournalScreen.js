import React from 'react';
import { Sidebar } from './Sidebar';
import { NothingSelect } from './NothingSelect';
import { NoteScreen } from '../notes/NoteScreen';
import { useSelector } from 'react-redux';

export const JournalScreen = () => {

    const { active } = useSelector( state => state.notes );

    return (
        <div className="journal__main-content animate__animated animate__fadeIn animate__faster">
            
            <Sidebar />

            <main>
                {
                    ( active )
                        ? ( <NoteScreen /> )
                        : ( <NothingSelect /> )
                }
            </main>
        </div>
    )
}
