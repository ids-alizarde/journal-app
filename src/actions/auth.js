import Swal from 'sweetalert2';
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { Constants } from '../services/constants';
import { noteLogoutClean } from './notes';
import { startLoading, finishLoading } from './ui';

export const startLoginEmailPassword = ( email, password ) => {
    return ( dispatch ) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword( email, password )
            .then( ({ user }) => {

                dispatch(Login( user.uid, user.displayName ));
                dispatch( finishLoading() );
            })
            .catch( e => {
                console.log(e);
                dispatch( finishLoading() );
                Swal.fire( 'Error', e.message, 'error' );
            });
    }
}

export const startRegisterWithEmailPasswordName = ( email, password, name ) => {
    return ( dispatch ) => {
        
        return firebase.auth().createUserWithEmailAndPassword( email, password )
            .then( async({ user }) => {
                await user.updateProfile({ displayName: name });
                dispatch(Login( user.uid, user.displayName ));
            })
            .catch( e => {
                // console.log(e);
                Swal.fire( 'Error', e.message, 'error' );
            });
            
    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch(Login( user.uid, user.displayName ));
            });
    }
}

export const Login = ( uid, displayName ) => ({
    type: Constants.login,
    payload: {
        uid,
        displayName
    }
});

export const startLogOut = () => {
    return async( dispatch ) => {

        await firebase.auth().signOut();
        dispatch( logOut() );
        dispatch( noteLogoutClean() ); 
    }
}

export const logOut = () => ({
    type: Constants.logout
});