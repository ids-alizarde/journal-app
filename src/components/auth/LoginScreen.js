import React from 'react';
import { Link } from 'react-router-dom'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { startLoginEmailPassword, startGoogleLogin } from '../../actions/auth';
import { setError, removeError } from '../../actions/ui';



export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { loading, msgError } = useSelector( state => state.UI )

    const [ formValues, handleInputChange ] = useForm({
        email: 'aldo@mail.com',
        password: '123456'
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();

        if ( isFormValid() ) {

            dispatch( startLoginEmailPassword( email, password ));
        }
    }

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    }

    const isFormValid = () => {

       if ( !validator.isEmail( email ) ) {

            dispatch( setError( 'Email isnÂ´t valid' ) );
            return false;
        } else if ( password.length < 5 ) {

            dispatch( setError( 'Password should be at least 6 characters and match each other' ) );
            return false;
        }

        dispatch( removeError() );
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form onSubmit={ handleLogin } 
                className="animate__animated animate__fadeIn animate__faster">
                { 
                    msgError !== null &&
                    (
                        <div className="auth__alert-error auth__animated auth__fadeInUp">
                            { msgError }
                        </div>
                    )
                }
                <input type="text"
                        className="auth__input"
                        placeholder="Em@il"
                        name="email" 
                        autoComplete="off"
                        value={ email }
                        onChange={ handleInputChange } />

                <input type="password"
                        className="auth__input"
                        placeholder="Password"
                        name="password" 
                        autoComplete="off"
                        value={ password }
                        onChange={ handleInputChange } />

                <button type="submit"
                        className="btn btn-block btn-primary"
                        disabled={ loading } >
                    Login
                </button>

                <div className="auth__social-network">
                    <p>Login with Google</p>
                    <div className="google-btn"
                            onClick={ handleGoogleLogin }>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to="/auth/register"
                        className="link">
                    Create a new account
                </Link>
            </form>
        </>
    )
}
