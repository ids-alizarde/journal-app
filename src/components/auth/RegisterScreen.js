import React from 'react';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { setError, removeError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';


export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { msgError } = useSelector( state => state.UI )

    const [ formValues, handleInputChange, reset ] = useForm({
        name: 'Aldo Ivan',
        email: 'aldo@mail.com',
        password: '123456',
        password2: '123456'

    });

    const { name, email, password, password2 } =  formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        
        if ( isFormValid() ) {
            dispatch( startRegisterWithEmailPasswordName( email, password, name) );
        }
    }

    const isFormValid = () => {

        if (name.trim().length === 0 ) {
            
            dispatch( setError( 'Name is required' ) );
            return false;
        } else if ( !validator.isEmail( email ) ) {

            dispatch( setError( 'Email isnÂ´t valid' ) );
            return false;
        } else if ( password !== password2 || password.length < 5 ) {

            dispatch( setError( 'Password should be at least 6 characters and match each other' ) );
            return false;
        }

        dispatch( removeError() );
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form onSubmit={ handleRegister }
                className="animate__animated animate__fadeIn animate__faster" >
                { 
                    msgError !== null &&
                    (
                        <div className="auth__alert-error auth__animated auth__fadeInUp">
                            { msgError }
                        </div>
                    )
                }
                <input type="text"
                        className="auth__input"
                        placeholder="Name"
                        name="name" 
                        autoComplete="off"
                        value={ name }
                        onChange={ handleInputChange } />

                <input type="text"
                        className="auth__input"
                        placeholder="Em@il"
                        name="email" 
                        autoComplete="off"
                        value={ email }
                        onChange={ handleInputChange } />

                <input type="password"
                        className="auth__input"
                        placeholder="Password"
                        name="password" 
                        autoComplete="off"
                        value={ password }
                        onChange={ handleInputChange } />
                
                <input type="password"
                        className="auth__input"
                        placeholder="Comfirm Password"
                        name="password2" 
                        autoComplete="off"
                        value={ password2 }
                        onChange={ handleInputChange } />

                <button type="submit"
                        className="btn btn-block btn-primary mb-5">
                    Register
                </button>

                <Link to="/auth/login"
                        className="link mt-5">
                    Already register ?
                </Link>
            </form>
        </>
    )
}
