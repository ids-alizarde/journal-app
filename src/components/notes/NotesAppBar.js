import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active: note } = useSelector( state => state.notes );

    const handleSaveNote = () => {

        dispatch( startSaveNote( note ) );
    }

    const handleUploadImage = () => {
        
        document.querySelector('#fileSelector').click();
    }

    const handleFileChange = (e) => {

        const file = e.target.files[0];

        if ( file ) {

            dispatch( startUploading( file ) );
        }
    }

    return (
        <div className="notes__appbar">
            <span>28 de agosto 2020</span>
            <input id="fileSelector"
                type="file"
                style={{ display: 'none' }}
                onChange={ handleFileChange } />
            <div>
                <button className="btn"
                    onClick={ handleUploadImage }>
                    Picture
                </button>
                <button className="btn"
                        onClick={ handleSaveNote }>
                    Save
                </button>
            </div>
        </div>
    )
}
