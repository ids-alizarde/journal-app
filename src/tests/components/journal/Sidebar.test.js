import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import { Constants } from '../../../services/constants';
import { act } from '@testing-library/react';
import { startLogOut } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';
import { Sidebar } from '../../../components/journal/Sidebar';

jest.mock('../../../actions/auth', () => ({

    startLogOut: jest.fn()
}));

jest.mock('../../../actions/notes', () => ({

    startNewNote: jest.fn()
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initialState = {
    auth: {
        uid: 'TESTING',
        name: 'Aldo Ivan'
    },
    UI: {
        loading: false,
        msgError: null
    },
    notes: {
        active: null,
        notes: []
    }
};

let store = mockStore( initialState );
store.dispatch = jest.fn();

const wrapper = mount( 
        <Provider store={ store }>
            <Sidebar/> 
        </Provider>
    );


describe('Pruebas en <Sidebar/>', () => {
    
    test('Debe de mostrarse correctamente el <Sidebar/>', () => {
        
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de llamar el startLogOut ', () => {
        
        wrapper.find( '.btn' ).prop( 'onClick' )();

        expect( startLogOut ).toHaveBeenCalled();
        expect( startLogOut ).toHaveBeenCalledTimes(1);
    });
    
    test('Debe de llamar el startNewNote', () => {
        
        wrapper.find( '.journal__new-entry' ).simulate( 'click' );

        expect( startNewNote ).toHaveBeenCalled();
        expect( startNewNote ).toHaveBeenCalledTimes(1);
    });
    
})
