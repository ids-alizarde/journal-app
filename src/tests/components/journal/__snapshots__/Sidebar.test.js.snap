// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Sidebar/> Debe de mostrarse correctamente el <Sidebar/> 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        aria-hidden="true"
        className="far fa-moon"
      />
      <span>
         
        Aldo Ivan
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries pointer"
  />
</aside>
`;
